nvflare/app_common/workflows/base_fedavg.py

```diff
@@ -0,0 +1,422 @@
+# Copyright (c) 2021, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import random
+from abc import ABC, abstractmethod
+from typing import List, Union
+
+from nvflare.apis.client import Client
+from nvflare.apis.controller_spec import OperatorMethod, TaskOperatorKey
+from nvflare.apis.fl_component import FLComponentHelper
+from nvflare.apis.fl_constant import FLMetaKey, ReturnCode
+from nvflare.apis.fl_context import FLContext
+from nvflare.apis.impl.controller import ClientTask, Controller, Task
+from nvflare.apis.shareable import Shareable
+from nvflare.apis.signal import Signal
+from nvflare.app_common.abstract.fl_model import FLModel, ParamsType
+from nvflare.app_common.abstract.learnable_persistor import LearnablePersistor
+from nvflare.app_common.abstract.model import ModelLearnableKey
+from nvflare.app_common.aggregators.weighted_aggregation_helper import WeightedAggregationHelper
+from nvflare.app_common.app_constant import AppConstants
+from nvflare.app_common.app_event_type import AppEventType
+from nvflare.app_common.utils.fl_model_utils import FLModelUtils
+from nvflare.security.logging import secure_format_exception
+from nvflare.widgets.info_collector import GroupInfoCollector, InfoCollector
+
+
+class FedAvgModelController(Controller, FLComponentHelper, ABC):
```

Comment:
Majority of the code is copied from SAG. I don't think this is the right approach.
We could refactor SAG to become more customizable, and then build this on top of SAG.

Reply:
Introduce `ModelController` as the base class.